### ISSUE TITLE
Allow for privacy friendly, cookie-less surfing of zotonic website

### ISSUE THREAD

**Issue by hce** [zotonic/zotonic#3826]
## `Zotonic` version
I'm using the latest github release, cca31849d3e86a5357f6b42aa261cc808cc3c0f5

## `OS` version
FreeBSD 14.1-RELEASE-p2

## Description
* **Motivation**

In some countries like Germany there are very strict laws in place with regards to cookie use, data protection etc. A simple solution to not having to deal with any of this might be to not use any cookies or other local storage on the website visiting user's browser. This makes sense especially for websites like blogs, whose purpose is to simply be read by users, no account required.

* **Proposal**

Add a setting to zotonic that allows to disable the use of all cookies, local and session storage as long as the zotonic website is browsed anonymously, i.e., as long as no-one is trying to log in.

## Current behavior

Currently a lot of cookies are set when visiting a zotonic website, both by the zotonic server as well as the javascript. Amongst those cookies there seem to be some that may be used for tracking purposes.

## Expected behavior

When the new setting disable_all_cookies is set to "true", no cookies, session or local storage would be set when visiting a zotonic powered site.

**Comment by mworrell** [comment]
The cookies that we use are all functional cookies and as such in compliance with the GDPR, they are:

 - `z.lang` language selection. This is for keeping the language consistent and is needed when switching between pages with and without language code.
 - `cotonic-sid`, this is needed to keep communication between tabs working and for some functionalities (especially when server-side data need to be stored), and security of the `z.auth` token. Currently as cookie and in localStorage.
 - `z.auth` session related - this is the authentication cookie. It also contains a session id. This is needed for security and communication.
 - `z.tz` the timezone as found out from the configs, the user preference, and the browser. We need this to display dates and times in the correct format.

And temporary data in sessionStorage. Which doesn't matter for the GDPR.

What might be possible is to make the `z.lang` and `z.tz` session cookies for anonymous users. The same could be done for the `cotonic-sid` cookie. The `cotonic-sid`  localStorage might be more problematic, but maybe that could be replaced with only the cookie.

For analytics/social/etc. types of cookies we have `mod_cookie_consent`

/cc @mmzeeman

**Comment by hce** [comment]
Thanks a lot for the quick response! So you are well aware of the GDPR. In this case I will just copy+paste what you have written above onto my blog's privacy statement and remove my nginx fiddling attempt at removing the cookies %} I didn't find any mention of the GDPR or cookies in the docs, did I just miss that bit or is it missing? ;)

Regardless of this I think supporting a "cookie-less" browsing experience is something you might want to consider supporting.

**Comment by mworrell** [comment]
Indeed, we didn't write down much in the docs themselves.  We have on some sites.

And yes, we are well aware of the GDPR, being based in the EU and such 😄

**Comment by hce** [comment]
Heh, yes, I know you are based in the EU :) I just thought since zotonic is a very technical project, maybe you're unaware or have just decided to implement a consensus popup per project etc., so I figured I'd ask ;-)

**Comment by mmzeeman** [comment]
For newer sites we usually don't use any analytics or social media cookies, so we don't have to ask permission. This will be presented when users register an account.

But I can see the benefit of having a simple... just-serve-pages-mode which does not setup a mqtt session via a web socket with every page visit. This would prevent having to set some ids in the sessionStorage. Do you have a specific use-case for this in mind?

Originally Zotonic was built to always have a bi-directional connection between the server and the browser. But it is not strictly necessary.

**Comment by hce** [comment]
My usecase is just to run a personal weblog, I could easily use a static site generator for that purpose but I like erlang and zotonic so I try to use it instead. My primary concern is complying with the GDPR and your clarification on which cookies are used and for what purposes and that it's GDPR compliant is sufficient for me and now I will keep using zotonic and not fiddle with the headers in nginx and not edit the JS files to not set any cookies. Thanks :)

What you say about always keeping the bi-directional connections open sounds like my suggestion is not really in line with zotonic's primary design goals and might unnecessarily complicate the project so I'd say just ignore the suggestion.

**Comment by mworrell** [comment]
One of the ideas is to make all cookies "session" cookies, unless someone logs on.
Or set a short keep-alive and extend when needed.

Then we would not leave traces when someone visited the site and the session data has been removed.

**Comment by hce** [comment]
Indeed, that would have a similar effect to removing the cookies, while keeping the advantages you mentioned. Using a short keep-alive might even be preferrable to using session cookies, as many users tend to keep their browsers open for ages these days, effectively never deleting any session cookies...

**Comment by mmzeeman** [comment]
FYI. Usually we setup Zotonic without having things like nginx in between. Erlang is more than capable enough to handle the connectivity. In the past we found passing all traffic through nginx basically doubled the response times.

Zotonic has a lot of built in functionality. People expect websites to have a lot of bells and whistles these days.

**Comment by hce** [comment]
Thanks, good to knokw. I'll keep that in mind. It doesn't really surprise me. Erlang is very good at handling concurrent network traffic and I guess that's especially true for long lasting connections. :-)

**Comment by mmzeeman** [comment]
Yes indeed... reverse proxies usually can't handle long lasting connections very well. And it always increases the amount of dependencies one needs to maintain... (I'm busy with a client which is using Envoy right now). We also have the feeling that erlangs ssl implementation is a bit more secure than something based on openssl.

**Comment by hce** [comment]
Oh wow, you mean erlang has got its own, native erlang TLS implementation? That's awesome! I had assumed it was just an openssl wrapper. Okay, you have convinced me to not use nginx for my zotonic blog. :)

**Comment by mworrell** [comment]
Yes, and with `mod_ssl_letsencrypt` you can request certificates and have them automatically extended.

**Comment by hce** [comment]
Makes sense. Oh, a letsencrypt plugin is very convenient! Another reason to use a reverse proxy gone. :)

**Comment by mmzeeman** [comment]
Maybe that is one of our design goals... Reduce the number of external things you need to setup separately to implement a website with bells and whistles.. Zotonic can also do incoming and outgoing email for instance.

**Comment by hce** [comment]
Awesome, I set it up the way you recommended, works like a charm, including the letsencrypt plugin!
Since we have already moved away from the original topic, could you explain what this port is for:
    zotonic  beam.smp     987 16  tcp4   *:26229               *:*
It seems to be open to the world; I was able to connect to it from a public IP. I didn't find a setting to disable it...

**Comment by mmzeeman** [comment]
Those are probably open sockets managed by `epmd` which controls how distributed erlang nodes connect to each other. You can control that with an environment variable.

See our cloud init file: https://github.com/zotonic/zotonic/blob/d18460e72dfb33df5ddddb6e75922e9a54eafa73/cloud-init/zotonic-cloudinit.yml#L60

With this setting `epmd` does not bind to the outside network interface, but only to the loopback device.

See also: https://www.erlang.org/docs/18/man/epmd

For more information on distributed erlang see: https://www.erlang.org/doc/system/distributed.html

**Comment by hce** [comment]
Hmm, I don't think that was it, because I'm getting:

    USER     COMMAND    PID   FD  PROTO  LOCAL ADDRESS         FOREIGN ADDRESS
    zotonic  epmd        8296 3   tcp4   127.0.0.1:4369        *:*

but also:

    USER     COMMAND    PID   FD  PROTO  LOCAL ADDRESS         FOREIGN ADDRESS
    zotonic  beam.smp    8298 16  tcp4   *:16467               *:*

The only module I've explicitly enabled is the letsencrypt one...

**Comment by mworrell** [comment]
MQTT, SMTP and HTTP(S) are definitely not on that port.
We don't add random listeners - so might have something to do with the Erlang distribution.

Maybe check the open ports in Erlang?

> recon:port_types().

And

> recon:tcp().

And

> recon:port_info(lists:nth(1, recon:tcp())).

And then for the Pid of the connected process (example Pid):

> recon:info("<0.64.0>").

I think it might actually be the Erlang distribution.
On my local dev machine the first TCP port has `{sockname,{{0,0,0,0},49225}}`
Which is connected to `<0.64.0>` which has the registered name `net_kernel`

**Comment by mmzeeman** [comment]
It is indeed for erlang distribution. https://www.erlang.org/doc/apps/kernel/net_kernel.html

**Comment by mworrell** [comment]
The kernel can be configured to only listen on a certain IP address.

You could try to add this to the `erlang.config` file.

```
{kernel, [
    {inet_dist_listen_options, [{ip, {127,0,0,1}}]},
     ...
```

**Comment by mworrell** [comment]
@mmzeeman Per default we start epmd for 127.0.0.1   (and .1.1 on Linux).  Shall we add the above to the default `erlang.config` ?

Of course need a some testing on Linux.

**Comment by mmzeeman** [comment]
Sounds like a plan. We usually don't use erlang distribution outside the machine anyway. So there is no need binding it to an external port.

**Comment by hce** [comment]
Thanks, that did it. I noticed a port bound to * during zotonic startup, but once startup is complete, all ports except 443 and 80 are now bound to 127.0.0.1 :-)

**Comment by mmzeeman** [comment]
It can also be configured via the command line.

> **Edit** Just tried on Linux (Ubuntu) and this was not working well there. This because of binding the name of the machine to 127.0.1.1

Shouldn't that be replaced with the static ip address of the host? (if the machine has one)

**Comment by mworrell** [comment]
Problem is that Linux adds two "loopback" addresses, for example, this is the default hosts file on the demo.zotonic.com server:

```
127.0.1.1 zotonic-demo zotonic-demo
127.0.0.1 localhost

# The following lines are desirable for IPv6 capable hosts
::1 localhost ip6-localhost ip6-loopback
ff02::1 ip6-allnodes
ff02::2 ip6-allrouters
```

And when we start Zotonic, it grabs the name of the host `zotonic-demo` and the erlang distribution will try to connect to that hostname.  So we need to listen on `{127,0,1,1}` for that one.

**Comment by mmzeeman** [comment]
Hmm... I always thought the .1.1 address was reserved for linux hosts without a static ip-address.

**Comment by mmzeeman** [comment]
This reminds me of a configuration error where somebody put `localhost` behind the static ip of the server. 😱

**Comment by mworrell** [comment]
So, it is a bit messy. We can actually do a lookup of the hostname being used and use that for the kernel distribution options.

```
1> inet_hosts:gethostbyname(inet_db:gethostname()).
{ok,{hostent,"zotonic-demo",
             ["zotonic-demo"],
             inet,4,
             [{127,0,1,1},{127,0,1,1}]}}
```

**Comment by mworrell** [comment]
> Hmm... I always thought the .1.1 address was reserved for linux hosts without a static ip-address. When you have a static ip you should replace 127.0.1.1 with your real ip. But I'm not a sysadmin... :-)

Yes, they say that you should edit that entry in case of a permanent IP address.
If you do that then our Erlang distribution will fail miserably if we let kernel listen on 127.0.0.1, as we (and Erlang) do use the hostname for the nodename when using short names.

**Comment by mmzeeman** [comment]
Or use the `net` module for hostname/address lookup? I'm always unsure what module to use for this.

**Comment by mworrell** [comment]
> Or use the net module for hostname/address lookup? I'm always unsure what module to use for this.

In `zotonic_command_nodename.erl` you can see how we (and Erlang) find the hostname and create the nodename.

**Comment by mmzeeman** [comment]
I was just looking into this before this, because I was wondering if we should enable the letsencrypt module for the status site. Which can be useful when you have multiple virtual host sites on the machine and want to send email from the site configured as the reverse-dns of the machine.

**Comment by mmzeeman** [comment]
It looks like `inet_db` is a not documented erlang api-module which should not be called from applications.

**Comment by mworrell** [comment]
We have some changes in the SNAME handling that should fix the issue of the distribution listening on public IP addresses.

https://github.com/zotonic/zotonic/pull/3829
